crates/shared/src/fee.rs

```diff
@@ -0,0 +1,68 @@
+use ethcontract::U256;
+
+/// Everything required to compute the fee amount in sell token
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct FeeParameters {
+    /// The estimated gas units required to execute the quoted trade.
+    pub gas_amount: f64,
+    /// The estimated gas price at the time of quoting.
+    pub gas_price: f64,
+    /// The Ether-denominated price of token at the time of quoting.
+    ///
+    /// The Ether value of `x` sell tokens is `x * sell_token_price`.
+    pub sell_token_price: f64,
+}
+
+impl Default for FeeParameters {
+    fn default() -> Self {
+        Self {
+            gas_amount: 0.,
+            gas_price: 0.,
+            // We can't use `derive(Default)` because then this field would have
+            // a value of `0.` and it is used in division. The actual value we
+            // use here doesn't really matter as long as its non-zero (since the
+            // resulting amount in native token or sell token will be 0
+            // regardless), but the multiplicative identity seemed like a
+            // natural default value to use.
+            sell_token_price: 1.,
+        }
+    }
+}
+
+impl FeeParameters {
+    pub fn fee(&self) -> U256 {
+        self.fee_with_additional_cost(0u64)
+    }
+
+    pub fn fee_with_additional_cost(&self, additional_cost: u64) -> U256 {
+        let fee_in_eth = (self.gas_amount + additional_cost as f64) * self.gas_price;
+
+        dtou(fee_in_eth / self.sell_token_price)
+    }
+}
+
+/// Converts an `f64` to a `U256`.
+///
+/// We want the conversion from f64 to U256 to use ceil because:
+/// 1. For final amounts that end up close to 0 atoms we always take a fee so we
+///    are not attackable through low decimal tokens.
+/// 2. When validating fees this consistently picks the same amount.
+fn dtou(d: f64) -> U256 {
+    U256::from_f64_lossy(d.ceil())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    // Convenience to allow using u32 in tests instead of the struct.
+    impl From<u32> for FeeParameters {
+        fn from(v: u32) -> Self {
+            FeeParameters {
+                gas_amount: v as f64,
+                gas_price: 1.0,
+                sell_token_price: 1.0,
+            }
+        }
+    }
+}
```

Comment:
As far as I can tell this is not used anywhere (thankfully so)

Reply:
indeed, removed 👍